NF: Mock Task manager

Arthur-Milchior opened the pull request:
Currently, the biggest problem to generate test are that some code is executed in background. Either I wait until the end of the background task, which is not easy to do, or I the test in OnPostExecute. In this case, I need to actually check whether onPostExecute is executed which lead back to first problem.

Worse, as show https://github.com/ankidroid/Anki-Android/pull/6978 there is difference in background task between machines.
The trouble, according to me, is that those are not difference I actually want to consider in tests. In the example linked above, it just means on some machine the counts will takes more time to be updated than on some other machines. That's not actually a problem as long as counts gets updated.

The trouble is that if I want to do test without task actually in background, currently, I need to run copy the background code in the test. This makes code complex, and potentially it does not even follow evolution of the task.


I thought that the solution should be:
* Uses a task manager. It contains what was static in CollectionTask
* Mock the task manager, with a method executing all tasks immediately. This way, this part of the test becomes deterministic. 
* Use the mock in the tests.

A single error remains, `checkDatabaseWithLockedCollectionReturnsLocked`. I'd love your help @david-allison-1 debugging this one. I don't know what this is supposed to test, so no way of knowing how to correct.
I tried to simply do  `getCol().getTaskManager().launchCollection(CHECK_DATABASE)` but this method never ends, so that's not the correct answer

david-allison-1 reviewed (commented):
This is a very wide-reaching change.

I think we're moving further to the point that we'll want to use dependency injection, but I don't think we're there quite yet.

I don't think the collection should contain the task scheduler.

I think the first stage would be to keep `CollectionTask.launchCollectionTask` and use a singleton that we can inject for testing. After that, there'll be a lot less code to review, and we can focus on potential implementation issues.

Arthur-Milchior commented:
I do not see why we would want to treat Time and TaskManager differently. So it seemed to me to make sens to have it also in the collection. I would have understood to have both is AnkiActivity, in CollectionHelper or somewhere else. But I would assume they should remain together since they have similar roles.

Actually, there should be no change at all appart in tests. The static variables are members of a class, but appart from that the behavior should be exactly the same hopefully. Anyway, it seems it's generally a bad taste to use static variables and methods, so encapsulating this in a class seems a good step in terme of usual java standard of good code.

Would you have a link explaining why injection would be better than giving the task manager at construction ? I do not know or see how it would improve things.


Where would you want your singleton ? I don't understand what you are asking for. Neither whether you agree with the general notion that, to get simpler tests, it would be a good idea not to have to deal with backgrounds

david-allison-1 commented:
Regarding the test failure, `execute` was previously called to obtain the result of the task to return it.

Now, this calls the internal execute method, because it wasn't provided with the correct argument.

https://github.com/ankidroid/Anki-Android/blob/95a35baf949ef169dd951f2a612fa1b572186a25/AnkiDroid/src/test/java/com/ichi2/async/AbstractCollectionTaskTest.java#L36

Changing it to `return task.execute(new TaskData[] { null }).get();` will work, but it's not great.

Arthur-Milchior commented:
Rebased over https://github.com/ankidroid/Anki-Android/pull/7006 to avoid creating useless objects

Arthur-Milchior commented:
> Regarding the test failure, `execute` was previously called to obtain the result of the task to return it.
> 
> Now, this calls the internal execute method, because it wasn't provided with the correct argument.
> 
> https://github.com/ankidroid/Anki-Android/blob/95a35baf949ef169dd951f2a612fa1b572186a25/AnkiDroid/src/test/java/com/ichi2/async/AbstractCollectionTaskTest.java#L36
> 
> Changing it to `return task.execute(new TaskData[] { null }).get();` will work, but it's not great.

The main problem I believe would be that this actually call execute, so I assume will lead to some background work, which I want to avoid. I guess in this case it makes sens.

Anyway, I believe that considering this very test is pointless until we know whether it's a correct way to do test. 
I believe it'll allow to write without needing to worry about execution, so I really like it

david-allison-1 commented:
The test mocks the database, calls a method on the collection, and asserts the results are correct given the mock. I don't see why that's pointless?

Arthur-Milchior commented:
I fear I was misconstrued.
It's useless to discuss the details of this specific test if you think that the whole PR should be done from scratch in a different way. If the PR gets closed and not merged, the fact that it fails is of little importance

david-allison-1 commented:
> I do not see why we would want to treat Time and TaskManager differently. So it seemed to me to make sens to have it also in the collection. I would have understood to have both is AnkiActivity, in CollectionHelper or somewhere else. But I would assume they should remain together since they have similar roles.

That's fair - let's extract both. They don't belong in the collection.

> Actually, there should be no change at all appart in tests. The static variables are members of a class, but appart from that the behavior should be exactly the same hopefully. Anyway, it seems it's generally a bad taste to use static variables and methods, so encapsulating this in a class seems a good step in terme of usual java standard of good code.

There is a massive amount of code that's been changed, this makes it harder to understand how the codebase is changing over time.

> Would you have a link explaining why injection would be better than giving the task manager at construction ? I do not know or see how it would improve things.

Dependency injection? As a basic overview, it offloads the decision of what dependencies to create to a different part of the code. It'll still use the constructor.

> Neither whether you agree with the general notion that, to get simpler tests, it would be a good idea not to have to deal with backgrounds

We need a method of detecting whether a task has completed or not, the count of executed tasks, and when the queue of tasks is empty or not. 

In some instances, we'll want this to be run in the background, and in other instances we'll want this to block the test until completion. Prescribing a different threading model for our tests means we're getting further away from testing how the code actually executes. 

> Where would you want your singleton ? I don't understand what you are asking for. 

Define a `getInstance()` on TaskManger returning the current instance, use that inside `launchCollectionTask`. This keeps the API consistent with what we have currently, massively reduces the amount of code changed, and lets us focus on the issue at hand.

This will be easy to remove in the future if we want to go with a different architectural pattern later on (for example, if a static variable doesn't mesh well with parallel test runs).

david-allison-1 commented on AnkiDroid/src/test/java/com/ichi2/libanki/sched/AbstractSchedTest.java:
```diff
@@ -96,7 +96,7 @@ public void testUndoResetsCardCountsToCorrectValue() throws InterruptedException
         sched.answerCard(cardBeforeUndo, EASE_3);
 
-        waitForTask(UNDO, 5000);
+        getCol().getTaskManager().launchCollectionTask(UNDO);
```
There's semantic meaning in the previous name (admittedly the timeout variable can go)

I feel the new code is harder to reason about

david-allison-1 reviewed (commented):
My main surprise is that this does not touch `waitForAsyncTasksToComplete`, which is (IMO) probably the biggest win, both in terms of correctness and performance from this PR.

Is there a reason why it wasn't touched?

Arthur-Milchior commented:
> > I do not see why we would want to treat Time and TaskManager differently. So it seemed to me to make sens to have it also in the collection. I would have understood to have both is AnkiActivity, in CollectionHelper or somewhere else. But I would assume they should remain together since they have similar roles.
> 
> That's fair - let's extract both. They don't belong in the collection. (or at least let's discuss this).

I'm fine by starting with resolving this simple question: where to put the Time object used in the program. If you want us to extract it, fine with me. I expect this will lead the way to the task manager later

Arthur-Milchior commented:
> (partial review)
> 
> My main surprise is that this does not touch `waitForAsyncTasksToComplete`, which is (IMO) probably the biggest win, both in terms of correctness and performance from this PR.
> 
> Is there a reason why it wasn't touched?

No reason at all. You're right, I should have removed it. Indeed, the whole purpose is to be able to write test without needing to use those wait. This means that i should be able to delete them already without trouble... Except in the case of Assume. In the case of Assume, I still need to tell `advanceRobolectricLooper` to execute

And there is a problem. Three test fails. `jsTime4ShouldBeBlankIfButtonUnavailable` fails, which I guess may make sens. I can't force JS to execute linearly as I force the remaining of the code...

And ReviewerTest's `testMultipleCards` fails. `answerCardOrdinalAsGood(reviewer, 2); // card 2 is shown` does not work. I don't understand why yet. I uses some Timber to follow the execution, and it seems `mCurrentCard` was correctly changed, so I have no clue what is going on... which is especially problematic, because this is exactly the kind of tests, which uses background tasks, and that I wanted to ensure not to fail

Arthur-Milchior commented:
The previous problem is solved. It was due to the fact that publishing progress used the standard activity task. It's corrected and directly calls the listener through the manager.

Two problems remains. 
I get 
> java.lang.NullPointerException: Cannot run onActivity since Activity has been destroyed already

 for reasons I do not understand; but I don't understand what the test is actually doing.

`jsTime4ShouldBeBlankIfButtonUnavailable` still fails. I was expecting it to be related to be related to the fact that it is in background, but it does not seems that just adding  `waitForAsyncTasksToComplete();` back does not solve this problem

Arthur-Milchior commented:
I ran an experiment. The branch https://github.com/Arthur-Milchior/Anki-Android/tree/regression_6898_3 merges the current branch with https://github.com/ankidroid/Anki-Android/pull/6978 . It ran on travis on https://travis-ci.org/github/Arthur-Milchior/Anki-Android/jobs/722604264

And the good news is that it that `testLrnQueueAfterUndo` does not fail. Which shows, as I expected, that the problem was that the test ran before the values were changed in the reviewer and not because the underlying logic failed. According to me, this seems a proof that the whole idea actually allows to create test that would be complex to create otherwise.

Arthur-Milchior commented:
Rebased. I added a mock task manager but decided not to use it. 
The previous experience shows that it helps create unit test (at least when no UI is involved), without having to guess where to put robotelectric function to advance, so it still nice to have it in. And what I hope is that it's also a step toward removing the deprecated AsyncTask, since the Task Manager is now abstract and can be implemented in different way.

david-allison-1 commented on AnkiDroid/src/androidTest/java/com/ichi2/anki/tests/ContentProviderTest.java:
```diff
@@ -795,9 +795,9 @@ public void testQueryNextCard() {
             col.reset();
             nextCard = sched.getCard();
-            TaskManager.waitToFinish();
+            TaskManager.getManager().waitToFinish();
```
Let's keep the `TaskManager` methods static, and call `.getManager()` inside the TaskManager methods, this will reduce the PR down to what matters

david-allison-1 reviewed (commented):
I feel this will be a high priority to get working now we have more contributors. 

GitHub Actions has made tests much more flaky, and code coverage is also nondeterministic.

david-allison-1 commented:
For the methods that are broken: would we be able to define our own JUnit Runner/annotations which will run certain tests under a given `TaskManager`?

david-allison-1 commented on AnkiDroid/src/test/java/com/ichi2/anki/RobolectricTest.java:
```diff
@@ -167,10 +172,31 @@ public void tearDown() {
+     * Ensure that each task in backgrounds are executed immediately instead of being queued.
+     * This may help debugging test without requiring to guess where `advanceRobolectricLooper` are needed.
+     */
+    public void noBackground() {
```
```suggestion
    public void runTasksInForeground() {
```

david-allison-1 commented on AnkiDroid/src/test/java/com/ichi2/anki/RobolectricTest.java:
```diff
@@ -167,10 +172,31 @@ public void tearDown() {
             //there is no onDestroy(), so call it here.
             Timber.uprootAll();
+
+            TaskManager.setTaskManager(new SingleTaskManager());
```
Shouldn't this be a call to `background`, or else the local variable won't be set

Arthur-Milchior replied:
Nice catch

david-allison-1 commented on AnkiDroid/src/test/java/com/ichi2/async/ForegroundTaskManager.java:
```diff
@@ -0,0 +1,104 @@
+        if (listener != null) {
+            listener.onPreExecute();
+        }
+        ResultBackground res = task.task(mColGetter.getCol(), new MockTaskManager<>(listener));
```
Is it possible to add a warning to exceptions thrown here (as Robolectric does) to warn that a new failure may have something to do with running in the foreground

Arthur-Milchior replied:
Done. Not sure about the exact message

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/async/TaskManager.java:
```diff
@@ -50,11 +48,18 @@ protected static void setLatestInstance(CollectionTask task) {
+
     public static <ProgressBackground, ResultBackground> CollectionTask<ProgressBackground, ProgressBackground, ResultBackground, ResultBackground> launchCollectionTask(CollectionTask.Task<ProgressBackground, ResultBackground> task) {
-        return launchCollectionTask(task, null);
+        return sTaskManager.launchCollectionTask(task);
```
⚠️ This is calling the static `launchCollectionTask`

david-allison-1 reviewed (changes requested):
As long as the changes to production are non-functional, this feels like it'll provide value and be ready to merge when we ensure that tests are still performing the same work, and that we have 0 failures.

Tests are failing currently, but this looks good from my perspective.

I do think the `background()` method will eventually need to be removed and modified to be a different deterministic scheduler, but we'll cross that bridge in a review cycle or so when we see what the deterministic test failures are.

david-allison-1 commented:
Tests are StackOverflowing

Arthur-Milchior commented:
Done

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:
```diff
@@ -906,7 +904,7 @@ protected void onActivityResult(int requestCode, int resultCode, Intent intent)
                 }
             } else if (resultCode == Reviewer.RESULT_ABORT_AND_SYNC) {
                 Timber.i("Obtained Abort and Sync result");
-                CollectionTask.waitForAllToFinish(4);
```
This should likely still be an abstraction on CollectionTask

Arthur-Milchior replied:
Why ?
My understanding is that CollectionTask deal with the individual tasks and the TaskManager with the queue of tasks. This is relevant to the queue.

david-allison-1 reviewed (commented):
My comments from yesterday still stand, nothing major.

This seems like it's too easy, we could do with running CI a few times to ensure there's no flakiness 

But overall, if it's only this, I'm happy. Main changes seem to be non functional, so it's test only and the tests pass

Arthur-Milchior reviewed (commented):
Sorry, I missed your message from yesterday. They are now taken into account

Arthur-Milchior commented:
> This seems like it's too easy, we could do with running CI a few times to ensure there's no flakiness
> 
> But overall, if it's only this, I'm happy. Main changes seem to be non functional, so it's test only and the tests pass

I beg your pardon but I've no idea what you mean here. What seems to be too easy?

Currently, there should be no change at all since runTasksInForeground is never called. I don't really see why you'd expect more flakiness here.

Did someone kept a list of flaky tests ? To see if it may be useful if those one specifically work better with foreground

david-allison-1 reviewed (commented):
My mistake, but it doesn't change my opinion, let's get this in regardless

Flaky test list: https://github.com/ankidroid/Anki-Android/issues/8064

david-allison-1 reviewed (approved):
LGTM

mikehardy commented on AnkiDroid/src/test/java/com/ichi2/anki/RobolectricTest.java:
```diff
@@ -167,10 +172,31 @@ public void tearDown() {
+    /**
+     * Set back the standard background process
+     */
+    public void background() {
```
David's suggestion here is good, the API should have mirror naming if it has mirror effect I think

Arthur-Milchior replied:
Sorry, I missed it

mikehardy commented on AnkiDroid/src/test/java/com/ichi2/anki/RobolectricTest.java:
```diff
@@ -77,7 +80,9 @@
-public class RobolectricTest {
+public class RobolectricTest implements CollectionGetter {
+
+    private static boolean mBackground = true;
```
I think I would actually go farther with this!
I would start with mBackground = false and in the RobolectricTest.setUp I would set the foreground task manager, with cleanup resetting it to foreground

I think unit tests wanting actual asynchronous testing are the rare case in other words, and the burden should be on them to call `runTasksInForeground` vs on regular code to call `runTasksInBackground` (note I picked David's name there

Arthur-Milchior replied:
I would like to merge this PR first.
I'm okay with doing what you want, but in a different PR. Because if we go your way, we will probably remove also `advanceRobolectricLooper` and similar functions, as they won't be relevant anymore, and this may lead to some complex tests to deal with, discussion about which tests should or should not be on foreground/background... 
So, I believe that the request you ask for will take a lot of discussion, and the discussion will be more complex because the current PR will be in the change list.

I'd note that, as far as I know, merging this PR is still a progress, whatever the decision we take for when we should use foreground and background in tests.

mikehardy replied:
Fair point - okay, in this goes, and then hopefully we can switch the default and delete manual looper advance from most tests

Arthur-Milchior replied:
> I think I would actually go farther with this!
> I would start with mBackground = false and in the RobolectricTest.setUp I would set the foreground task manager, with cleanup resetting it to foreground

to "background" you mean I assume

mikehardy reviewed (changes requested):
I think this should help the test situation a lot, so much that my main comment is that I want it to actually go *farther* :-)

mikehardy commented:
@Arthur-Milchior compile error in here directly related to my request for an API name change, looks like you changed the name name in definition but not where you call it

Arthur-Milchior commented:
I guess I was more exhausted than I thought
